Don't show checkout block for logged out users unless store is configured to allow guest checkout

haszari opened the pull request:
Fixes #2252

Fixes #2349

There are a few existing options affecting checkout when user is logged out - this PR handles `Allow customers to place orders without an account`. If this is not enabled, then the checkout is only displayed when users are logged in.

This PR also has a fix for #2349 - easy to fix here since it also needs `CURRENT_USER_LOGGED_IN`. 

See also related issue #2325.

### Screenshots

<img width="524" alt="Screen Shot 2020-05-04 at 11 21 35 AM" src="https://user-images.githubusercontent.com/4167300/80928549-73f48800-8df9-11ea-9620-5bbc8961bd91.png">


### How to test the changes 
#### Test #2252 - guest checkout store option works correctly for checkout block
1. Set up a page with checkout block. 
2. Go to `WooCommerce > Settings > Accounts & Privacy` and toggle `Allow customers to place orders without an account`.
3. View checkout page when logged in and logged out / anon (use incognito window or other browser. Ensure that checkout is not available when logged out as per your store settings.

Compare with existing checkout. Note that the checkout shortcode allows user to sign up as part of checkout - so that behaviour is different (see [comment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2252#issuecomment-622208649) and new issue #2377).

#### Test #2349 - checkout should only show login prompt when appropriate and necessary
1. Go to `WooCommerce > Settings > Accounts & Privacy` and enable `Allow customers to log into an existing account during checkout`.
2. Add products to cart & view checkout when logged in and anon. Should show login prompt as appropriate.

mikejolley commented:
I also found 2349 and fixed it in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2374, but I didn't need to introduce global settings because we already expose the customerId via the CheckoutProvider.

nerrad commented:
I appreciate picking up restricting guest checkout in this pull (to respect the checkout setting), however I think in those cases we should include a link to login when the checkout is restricted.

haszari commented:
> I also found 2349 and fixed it in #2374, but I didn't need to introduce global settings because we already expose the customerId via the CheckoutProvider.

Aha - nice @mikejolley. I've reworked this PR to use that same approach for now (so this PR is self-consistent), I'm happy to rebase or remove this if needed :)

> I appreciate picking up restricting guest checkout in this pull (to respect the checkout setting), however I think in those cases we should include a link to login when the checkout is restricted.

Good point @nerrad  – I've added this now, see new screenshot in PR description.

nerrad commented:
Just a note, we should probably set the `redirect_to` parameter on the url you use for the link so that the shopper is returned to the checkout after logging in. We should probably do the same for the `loginPrompt` variable in `checkout/block.js` as well.

haszari commented:
>  Just a note, we should probably set the redirect_to parameter on the url you use for the link so that the shopper is returned to the checkout after logging in. We should probably do the same for the loginPrompt variable in checkout/block.js as well.

That's good feedback @nerrad – I think I'll get that sorted in this PR (for both) rather than add a follow up.
